### PR TITLE
[Refactor] opt auto casting of TfSavedModelArtifact & clearer feedback

### PR DESCRIPTION
**Pull request by bojiang**
## Description

When packing a model built with the `tf.keras` API:

```
[2020-11-13 19:06:52,091] WARNING - Due to TensorFlow's save and load mechanism, only methods marked by `tf.function` (and some keras defaults) could be restored after packing & loading.
You can test the restored model object by referring:
<bento_svc>.artifacts.model

[2020-11-13 19:06:52,093] INFO - Found restored functions:
model(inputs, training, mask)
------------
Arguments Option 1:
  Positional arguments (3 total):
    * TensorSpec(shape=(None, None), dtype=tf.float32, name='inputs')
    * False
    * None
  Keyword arguments:
    {}

Arguments Option 2:
  Positional arguments (3 total):
    * TensorSpec(shape=(None, None), dtype=tf.float32, name='embedding_input')
    * True
    * None
  Keyword arguments:
    {}

Arguments Option 3:
  Positional arguments (3 total):
    * TensorSpec(shape=(None, None), dtype=tf.float32, name='inputs')
    * True
    * None
  Keyword arguments:
    {}

Arguments Option 4:
  Positional arguments (3 total):
    * TensorSpec(shape=(None, None), dtype=tf.float32, name='embedding_input')
    * False
    * None
  Keyword arguments:
    {}

Return:
  TensorSpec(shape=(None, 2), dtype=tf.float32, name=None)


model.signature['serving_default'](embedding_input)
------------
Arguments Option 1:
  Positional arguments (0 total):
    * 
  Keyword arguments:
    {'embedding_input': TensorSpec(shape=(None, None), dtype=tf.float32, name='embedding_input')}

Return:
  {'dense_1': TensorSpec(shape=(None, 2), dtype=tf.float32, name='dense_1')}



[2020-11-13 19:06:52,093] WARNING - You may be using `TensorflowSavedModelArtifact` to pack a keras model. It works but may cause performance issues. Consider:
 * replacing `TensorflowSavedModelArtifact` with `KerasModelArtifact` (relatively good performance)
 * keeping `TensorflowSavedModelArtifact`, wrapping the `keras_model.predict` with a `tf.function`. (best performance)
```

## How Has This Been Tested?

- [x] tf2 + native API: integration test
- [x] tf2 + keras API: integration test
- [x] tf1 + native API (non-eager): tested manually on gallery fashion mnist example
- [ ] tf1 + estimator API: none

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [x] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [ ] BentoService (service definition, dependency management, API input/output adapters)
- [x] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/frameworks/tensorflow.py [hunk 1]
```diff
@@ -15,79 +22,69 @@ def _is_path_like(p):
...
+        if self.arg_specs is None and self.kwarg_specs is None:
             return self.origin_func(*args, **kwargs)
 
         for k in kwargs:
-            if k not in self._args_to_indices:
+            if k not in self.kwarg_specs:
                 raise TypeError(f"Function got an unexpected keyword argument {k}")
-        signatures_by_kw = {k: signatures[self._args_to_indices[k]] for k in kwargs}
+
+        arg_keys = {k for k, _ in zip(self.arg_names, args)}
```
Comment (parano) [inline bentoml/frameworks/tensorflow.py:46]: is this a convention in TensorFlow? any reason not to limit this function to keyword argument only?

Comment (bojiang) [reply bentoml/frameworks/tensorflow.py:46]: This a convention in Python lol

Comment (bojiang) [reply bentoml/frameworks/tensorflow.py:46]: In some special cases, the restored function accepts args and its `arg_names` is None.

Comment (parano) [reply bentoml/frameworks/tensorflow.py:46]: I mean, using the order of `arg_names`  to map positional argument of `__call__` back to the argument key

Comment (parano) [reply bentoml/frameworks/tensorflow.py:46]: This behavior seems very fragile to me.  It's good that you added a validation here, although from user perspective, it seems very hard to figure out the sequence of the arguments.  I might be missing some context here tho, that's why I ask if this is a convention in Tensorflow

File: bentoml/frameworks/tensorflow.py [hunk 2]
```diff
@@ -246,23 +240,39 @@ def pack(
...
                 )
             self._path = self._tmpdir.name
         else:
             self._path = obj
-
+        self._packed = True
+        loaded = self.get()
+        logger.warning(
+            f"Due to TensorFlow's save and load mechanism, only methods marked by "
+            f"`tf.function` (and some keras defaults) could be restored after "
```
Comment (parano) [inline bentoml/frameworks/tensorflow.py:253]: "Due to TensorFlow's internal mechanism, only methods annotated with `@tf.function` and Keras model API methods including `predict` and `xxxx` can be restored after a save & load."

File: bentoml/frameworks/tensorflow.py [hunk 3]
```diff
@@ -246,23 +240,39 @@ def pack(
...
+            f"packing & loading.\n"
+            f"You can test the restored model object by referring:\n"
+            f"<bento_svc>.artifacts.{self.name}\n"
+        )
+        logger.info(pretty_format_restored_model(loaded))
+        if hasattr(loaded, "keras_api"):
+            logger.warning(
+                f"You may be using `{self.__class__.__name__}` to pack a keras model. "
```
Comment (parano) [inline bentoml/frameworks/tensorflow.py:261]: `{self.__class__.__name__}` is being used to pack a TF model built with Keras API. In order to get optimal serving performance, we recommend either replacing `{self.__class__.__name__}` with `KerasModelArtifact`, or wrapping the `keras_model.predict` method with `tf.function` decorator.

File: bentoml/utils/tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    serving_default = get_serving_default_function(model)
+    if serving_default:
+        part_functions += pretty_format_function(
+            serving_default, "model", "signature['serving_default']"
+        )
+        part_functions += "\n"
+
+    if not restored_functions and not serving_default:
+        return (
+            "Found no avaliable functions. You can use `tf.function` "
```
Comment (parano) [inline bentoml/utils/tensorflow.py:209]: "No serving function was found in the saved model. In the model implementation, use `tf.function` decorator to mark the method needed for model serving. Find more details in related TensorFlow docs here https://www.tensorflow.org/api_docs/python/tf/saved_model/save"

**Pull request merged**